src/test/func/malloc/malloc.cc

```diff
@@ -138,6 +138,22 @@ void test_memalign(size_t size, size_t align, int err, bool null)
   check_result(size, align, p, err, null);
 }
 
+#if !defined(SNMALLOC_NO_REALLOCARR)
+void test_reallocarr(void* p, size_t nmemb, size_t size, int err, bool null)
+{
+  size_t old_size = 0;
+  if (p != nullptr)
+    old_size = our_malloc_usable_size(p);
+
+  printf("reallocarr(%p(%zu), %zu)\n", p, old_size, nmemb * size);
+  errno = SUCCESS;
+  int r = our_reallocarr(&p, nmemb, size);
```

Comment:
I think it would be good to store a bit pattern in the `reallocarr`, so that we can test we have actually copied the contents.  I liked what @schrodingerzhu did in his current PR for this on Rust.

https://github.com/microsoft/snmalloc/pull/426/files#diff-53e72133956e7966c55e6991080b396e2764f36321e757d4e2cee72bb959530bR17-R27

I think this would catch some bugs in your current code.

Reply:
For instance, I just added 
```C++
    size_t* s = (size_t*)our_malloc(16);
    *s = 123456;
    if (SUCCESS == our_reallocarr(&s, 16, 16))
    {
      if (*s != 123456)
      {
        puts("reallocarr: bad value");
        fflush(stdout);
        abort();
      }
    }
    else
    {
      our_free(p);
    }
```
and it prints bad value.  I think the issue in your current code is either use `realloc` and attempt to get the correct error codes out, or don't use `realloc` and expand the copying.

I would probably lift the end of `realloc` and rewrite something like:
```C++
    void* p = a.alloc(size);
    if (SNMALLOC_LIKELY(p == nullptr))
    {
       return ENOMEM;
    }

    sz = bits::min(size, sz);
    // Guard memcpy as GCC is assuming not nullptr for ptr after the memcpy
    // otherwise.
    if (sz != 0)
      memcpy(p, *ptr, sz);
    a.dealloc(*ptr);
    *ptr = p;
    return SUCCESS;
```